pkg/obfuscate/appsec.go

```diff
@@ -0,0 +1,538 @@
+	// is no sensitive key (which acts as a regexp obfuscating everything)
+	if o.valueRE == nil && !hasSensitiveKey {
+		return
+	}
+	from, to, err := scanString(input)
+	if err != nil {
+		return
+	}
+	if hasSensitiveKey {
+		diff.add(from, to-1, `"?"`)
+		return
+	}
+	value := input[from:to]
+	value, ok := unquote(value)
+	if !ok {
+		return
+	}
+	if !o.valueRE.MatchString(value) {
+		return
+	}
+	value = o.valueRE.ReplaceAllString(value, "?")
+	value, err = quote(value)
+	if err != nil {
+		return
+	}
+	diff.add(from, to-1, value)
+}
+
+// Helper function to walk the array elements of type string.
+func walkArrayStrings(input string, visit func(from int, to int)) {
+	var scanner scanner
+	scanner.reset()
+	i, err := stepTo(&scanner, input, 0, scanBeginArray)
+	if err != nil {
+		return
+	}
+	stringFrom := -1
+	depth := 0
+	for ; i < len(input); i++ {
+		c := input[i]
+		switch scanner.step(&scanner, c) {
+		case scanBeginObject, scanBeginArray:
+			depth++
+		case scanEndObject:
+			depth--
+		case scanEndArray:
+			if depth == 0 {
+				return
+			}
+			depth--
+
+		case scanBeginLiteral:
+			if depth == 0 && input[i] == '"' {
+				stringFrom = i
+			}
+		case scanContinue:
+			if stringFrom != -1 && input[i] == '"' && input[i-1] != '\\' {
+				visit(stringFrom, i)
+				stringFrom = -1
+			}
+
+		case scanError:
+			return
+		}
+	}
+}
+
+// Helper function to walk the object keys and values.
+func walkObject(scanner *scanner, input string, i int, visit func(keyFrom, keyTo, valueFrom, valueTo int)) (int, error) {
+	i, err := stepTo(scanner, input, i, scanBeginObject)
+	if err != nil {
+		return i, err
+	}
+	keyFrom := -1
+	keyTo := -1
+	valueFrom := -1
+	depth := 0
+	for ; i < len(input); i++ {
+		switch scanner.step(scanner, input[i]) {
+		case scanBeginObject, scanBeginArray:
+			depth++
+		case scanEndArray:
+			depth--
+		case scanEndObject:
+			if depth != 0 {
+				depth--
+				continue
+			}
+			// We reached the end of the object we were scanning
+			if keyFrom != -1 && keyTo != -1 && valueFrom != -1 {
+				// Visit the last value of the objet
+				visit(keyFrom, keyTo, valueFrom, i)
+			}
+			return i + 1, nil
+
+		case scanBeginLiteral:
+			if depth != 0 || keyFrom != -1 {
+				continue
+			}
+			if input[i] == '"' {
+				keyFrom = i
+			}
+		case scanContinue:
+			if keyFrom != -1 && keyTo == -1 && input[i] == '"' && input[i-1] != '\\' {
+				keyTo = i + 1
+			}
+
+		case scanObjectKey:
+			if depth == 0 {
+				valueFrom = i + 1
+			}
+		case scanObjectValue:
+			if depth != 0 {
+				continue
+			}
+			visit(keyFrom, keyTo, valueFrom, i)
+			keyFrom = -1
+			keyTo = -1
+			valueFrom = -1
+
+		case scanError:
+			return i, scanner.err
+		}
+	}
+	scanner.eof()
+	return i, scanner.err
+}
+
+// Helper function to step to the given expected scanner operation `to`.
+func stepTo(scanner *scanner, input string, i int, to int) (int, error) {
+	for ; i < len(input); i++ {
+		switch op := scanner.step(scanner, input[i]); op {
+		default:
```

Comment:
The `default` case always goes last. Does this arrangement provide any functional value in your case?

Reply:
No, no specific reason.
I moved it to last in https://github.com/DataDog/datadog-agent/pull/11324/commits/48234cfc628802b9faf4ed9289d9742eaf5dc83d